### PR TITLE
Upgrading from Unit overview improved

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
An old branch, 2 or 3 months old, revisited. I can't find its history here...

Anyway, it started out as more info on upgrades as tooltips, and the one on the UnitAction stayed more or less the same as back then - non-mousers can do without the extra info, as it's no so far away to find out.
<details><summary>screenshot</summary>

image
</details>
... positioning, colour choice, scale, or cutting it entirely are open to suggestions.

##
The unit overview change, however, is now completely different - a mini-menu on tap/click, which is equal rights for all, mobile or desktop:
<details><summary>mp4</summary>

https://github.com/yairm210/Unciv/assets/63000004/05b49466-d8a1-4c13-b088-e07083567fd6

</details>

... pretty ludicrous, over the top animation, but fun. Some architectural changes could be placed elsewhere - the way I abuse Popup could be encapsulated, or the ability to do smaller buttons or use Gdx's own button disabling instead of our kludge could be global. Key bindings, too - I chose to bind 'confirm' to do an upgrade, as the UnitAction bindings are still queued.

### CODE REVIEW

**Comment by AdityaMH** [comment]
> screenshot
> image

Right mouse/hold screen?

**Comment by yairm210** [comment]
Is the second one properly clickable on Android phones with my fat fingers?

**Comment by yairm210** [comment]
Looks great!

File: core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt [hunk 1]
```diff
@@ -304,4 +327,141 @@ class UnitOverviewTab(
         button.addAction(blinkAction)
         return scrollY
     }
+
+    private class UnitUpgradeMenu(
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt:331]: This file is to big as it is, please move this to a different file :)

File: core/src/com/unciv/ui/screens/mapeditorscreen/tabs/MapEditorEditSubTabs.kt [hunk 2]
```diff
@@ -244,7 +244,7 @@ class MapEditorEditImprovementsTab(
             }
             yield (FormattedLine(name, name, "Improvement/$name", size = 32))
         }
-    }.toList()
+    }.asIterable()
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/mapeditorscreen/tabs/MapEditorEditSubTabs.kt:247]: Isn't a sequence always an iterable...?

File: core/src/com/unciv/ui/objectdescriptions/BaseUnitDescriptions.kt [hunk 3]
```diff
@@ -231,4 +240,84 @@ object BaseUnitDescriptions {
...
+     * Lists differences e.g. for help on an upgrade, or how a nation-unique compares to its replacement.
+     *
+     * Cost is **not** included.
+     * Result lines are **not** translated.
+     *
+     * @param originalUnit The "older" unit
+     * @param betterUnit The "newer" unit
+     * @return Sequence of Pairs - first is the actual text, second is an optional link for Civilopedia use
+     */
+    fun getDifferences(ruleset: Ruleset, originalUnit: BaseUnit, betterUnit: BaseUnit):
```
Comment (yairm210) [inline core/src/com/unciv/ui/objectdescriptions/BaseUnitDescriptions.kt:254]: Nice!

File: core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt [hunk 4]
```diff
@@ -304,4 +327,141 @@ class UnitOverviewTab(
...
+                upgradeAllButton.isDisabled = unit.civ.gold < allCost ||
+                    allResources.isNotEmpty() &&
+                    unit.civ.getCivResourcesByName().run {
+                        allResources.any {
+                            it.value > (this[it.key] ?: 0)
+                        }
+                    }
+                upgradeAllButton.onActivation(::doAllUpgrade)
+                newInnerTable.add(upgradeAllButton).pad(2f, 15f).growX().row()
+            }
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt:384]: Separate to 'tryAddUpgradeAllUnitsButton' function

File: core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt [hunk 5]
```diff
@@ -304,4 +327,141 @@ class UnitOverviewTab(
...
+
+            clickBehindToClose = true
+            keyShortcuts.add(KeyCharAndCode.BACK) { close() }
+
+            newInnerTable.pack()
+            container = Container(newInnerTable)
+            container.touchable = Touchable.childrenOnly
+            container.isTransform = true
+            container.setScale(0.05f)
+            container.color.a = 0f
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/overviewscreen/UnitOverviewTab.kt:394]: Oh niiiice, that's how you did the animation! Cool beans

**Comment by Ouaz** [comment]
image

A late and nitpicky comment but `U: Upgrade` looks weird when translated in another language, as this is not the same initial (in french: `U: Améliorer`)
Why not simply display "`Upgrade to [new_unit_icon]`"? (move the unit icon to the right, and remove the the red `U:`)

tooltip

**Comment by WhoIsJohannes** [comment]
Thanks for making this change. It's really cool! The upgrade overlay sometimes gets stuck for me. Any idea why?

**Comment by WhoIsJohannes** [comment]
Screenshot_20230604-121415
Screenshot_20230604-121205

**Comment by WhoIsJohannes** [comment]
Maybe it was a hickup. Maybe we can ignore it unless many people report it.
Otherwise we need it to be reproducible, I agree.

Il dom 4 giu 2023, 16:39 SomeTroglodyte ***@***.***> ha
scritto:

**Comment by WhoIsJohannes** [comment]
Here is a repro when you upgrade to the mech infantry on the fifth or so unit. Sorry for the drive link, no idea how to add it otherwise on my phone.

https://drive.google.com/file/d/1jK7TKWHmUh4b0H11lH4NsFDWGOJzEa7J/view?usp=drivesdk

**Comment by WhoIsJohannes** [comment]
So this is on mobile and I'm upgrading from the unit menu on the map screen. It seems like it's coming out after the tap, but then doesn't disappear with the menu. Was this supposed to be a hover action on desktop and you didn't consider mobile maybe? At least that's how it seems to me.

**Pull request merged**